---
title: "methylation_QC"
author: "Marta Perez Alcantara"
date: "3 de octubre de 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Methylation EPIC (850k) QC

```{r load}
library(ChAMP)
library(RColorBrewer)
library(minfi)
library(ENmix)
library(GEOquery)
library(geneplotter)
library(reshape2)
library(gridExtra)


`%notin%` <- function(x,y) !(x %in% y)

# Cross-reactive probes
cross_reactive_EPIC <- read.csv("/Users/Marta/Documents/WTCHG/DPhil/Data/Regulation/Methylation/cross-reactive_probes_Pidsley_2016/EPIC_cross_reactive_probes.csv",sep=";")

baseDir = "/Users/Marta/Documents/WTCHG/DPhil/Data/Regulation/Methylation/P160281_MethylationEPIC_NicolaBeer/03.archive/P160281_MethylationEPIC_NicolaBeer.all.tar"
rgSet = readidat(path = baseDir, recursive = T)
sheet = read.metharray.sheet(base="/Users/Marta/Documents/WTCHG/DPhil/Data/Regulation/Methylation/P160281_MethylationEPIC_NicolaBeer/03.archive/3.Results", recursive = F)


# add stage column

sheet$stage= c("iPSC","iPSC","iPSC","DE","DE","DE","GT","GT","GT","PF","PF","PF","PE","PE","PE","EP","EN","EN","EN","BLC","BLC","124I","141 (14-19)","177","179","182","184","124R","136","137","138","139")

# add sample column
sheet$sample=c("SBNeo1.1","SBAd2.1","SBAd3.1","SBNeo1.1","SBAd2.1","SBAd3.1","SBNeo1.1","SBAd2.1","SBAd3.1","SBNeo1.1","SBAd2.1","SBAd3.1","SBNeo1.1","SBAd2.1","SBAd3.1","SBAd3.1","SBNeo1.1","SBAd2.1","SBAd3.1","SBNeo1.1","SBAd3.1","ISL","ISL","ISL","ISL","ISL","ISL","R","R","R","R","R")

sheet$Sample_Name=gsub( "^.*?_", "", sheet$Sample_Name) # take out stage part of experiment id (sample name)
rownames(sheet) <- sheet$Sample_Name # add that as rownames
colData(rgSet)=as(sheet[colnames(rgSet),],"DataFrame")

```

rgSet is a green/red channel object, that contains the raw intensities of the probes. I haven't normalised anything yet. We see that the samples are named following a standard IDAT naming convention with a 10 digit number, which is an array identifier followed by an identifier of the form R01C01.Example: The 200526570053_R01C01 means row 1 and column 1 on chip 200526570053.

I now get the standard GEO representation to get the phenotype data stored in GEO. Most of the columns in this phenotype data are irrelevant (contains data such as the person who submitted the data).


Now I try parts of the ENmix tutorial. I use a modified version of plotCtrl to detect anormalities in control probes. There is a compact version of this in QC_probes_plots.R
```{r ENmix_probes_QC}


 plotCtrl(rgSet)  # generates all probes plots. Check with Illumina manual to detect anormalities.

```
```{QC_type}
# transform to methylset

#This function matches up the different probes and color channels. Note that the dimension of this object is much smaller than for the RGChannelSet; this is because CpGs measured by type I probes are measured by 2 probes.

# The accessors getMeth and getUnmeth can be used to get the methylated and unmethylated intensities matrices:


mraw = getmeth(rgSet)


##########modifying multifreqpoly###########

bincount <- function(x,breaks){
  x <- x[!is.na(x)]
  bc <- table(.bincode(x, breaks, TRUE, TRUE))
  temp=data.frame(id=c(1: (length(breaks)-1)))
  bc=data.frame(id=as.numeric(names(bc)),counts=as.numeric(bc))
  resu=merge(temp,bc,by="id",all.x=TRUE)
  resu$counts[is.na(resu$counts)]=0
  resu$counts[order(resu$id)]
}

multifreqpoly <- function(mat, nbreaks=100, col=1:ncol(mat), xlab="Beta", renaming_sheet=sheet,
                          ylab="Frequency", main = .main, legend = list(x = "topright", fill=col,inset=c(-0.2,0),cex=0.6,
                                                          legend = if(is.null(colnames(mat))) paste(1:ncol(mat)) else
                                                            colnames(mat)),...)
{
  if(!is.matrix(mat)) stop("Warning: input data is not a numeric matrix\n")
  if(is.null(col)) col="black"
  col=rep(col,ceiling(ncol(mat)/length(col)))
  if(nbreaks > nrow(mat)) nbreaks=min(15,round(nrow(mat)/2))

  breaks <- seq(min(mat,na.rm=TRUE), max(mat,na.rm=TRUE),
                diff(range(mat,na.rm=TRUE))/nbreaks)
  mids <- 0.5 * (breaks[-1] + breaks[-length(breaks)])
  counts <- sapply(data.frame(mat),bincount,breaks=breaks)
  df = as.data.frame(cbind(counts,breaks[2:length(breaks)]))
  # plot(range(mids),c(0,max(counts)),type="n",xlab=xlab,ylab=ylab,...)
  melted = melt(df,id.vars=rev(names(df))[1])
  colnames(melted)=c("Breaks","Samples","Frequency")
  melted$Samples = sub('.', '', melted$Samples)
  subdf = sheet[c("Sample_Name","stage")]
  colnames(subdf) = c("Samples","Stages")
  melted$Stages = subdf[match(melted$Samples, subdf$Samples),"Stages"]
  melted$Stages = factor(melted$Stages,levels = unique(subdf$Stages),ordered = T)
  # for(i in 1:ncol(counts)){lines(mids,counts[,i],col=col[i],...)}
  # if(is.list(legend)) do.call(graphics::legend, legend)
  p = ggplot(melted,aes(x=Breaks,y=Frequency, group = Samples, col = Stages)) +
    geom_line() +
    xlab(xlab) +
    ggtitle(main)
  return(p)
}
#####################

#frequency plots of beta values

beta <- getB(mraw)
anno=rowData(mraw)


beta1 = beta[anno$Infinium_Design_Type == "I", ]
beta2 = beta[anno$Infinium_Design_Type == "II", ]



p1 = multifreqpoly(beta,main="All probes",xlab="Beta value")

p2 = multifreqpoly(beta1,main="Infinium I probes",xlab="Beta value")

p3 = multifreqpoly(beta2,main="Infinium II probes",xlab="Beta value")
 

png("multifreq_plots_beta_values_methylset.png",height=6,width=18,units = "in",res = 600,type = "cairo")

print(grid.arrange(p1, p2, p3, nrow = 1))
dev.off()


# Just differentiation samples
p1 = multifreqpoly(beta[,c(sheet$Sample_Name[1:21])],main="All probes",xlab="Beta value")

p2 = multifreqpoly(beta1[,c(sheet$Sample_Name[1:18])],main="Infinium I probes",xlab="Beta value")

p3 = multifreqpoly(beta2[,c(sheet$Sample_Name[1:18])],main="Infinium II probes",xlab="Beta value")
 

png("multifreq_plots_beta_values_methylset_diffStagesOnly.png",height=6,width=18,units = "in",res = 600,type = "cairo")

print(grid.arrange(p1, p2, p3, nrow = 1))
dev.off()



#  Data quality measures, including detection P values, number of beads for each methylation read
#  and average intensities for bisulfite conversion probes can be extracted using the function QCinfo
#  from an object of RGChannelSetExtended. According default or user specified quality score
#  thresholds, the QCinfo can also identify and export a list of low quality samples and CpG probes.
#  Outlier samples in total intensity or beta value distribution were often excluded before further
#  analysis. Such samples were tricky to be identified, by default the argument outlier=TRUE will
#  trigger the function to identify these outlier samples automatically. Quality score figures from
#  QCinfo can be used to guide the selection of quality score thresholds. Low quality samples and
#  probes can be filtered out using QCfilter or preprocessENmix.


# CpGthre=0.01 implies that probes that fail for at least 1 sample are removed
  qc<-QCinfo(rgSet,samplethre=0.01,detPthre=0.000001,CpGthre=0.01)
#  #exclude before backgroud correction
 # mdat<-preprocessENmix(rgSet, bgParaEst="oob", dyeCorr="RELIC", QCinfo=qc, nCores=8)
  #this step causes weird beta values in multifreqpoly plots
 mdat=mraw
# remove probes from sex chromosomes
anno=rowData(mdat)

discard <- intersect(mdat@NAMES, anno$Name[anno$chr %in% c("chrX","chrY")])
length(discard) # 19681
(length(discard)/length(mdat@NAMES))*100 # to lose 18986 probes (2.3%)

# Drop cross-hybridizing probes (aprox 5%) from Pidsley et al., 2016

probeNames <- mdat@NAMES
discard2 <- intersect(probeNames,as.character(cross_reactive_EPIC[,1])) # keep only probes that are not cross reactive
length(discard2) # 42847
(length(discard2)/length(mdat@NAMES))*100   #5%

discard=unique(c(discard,discard2)) # 60523
(length(discard)/length(mdat@NAMES))*100  # 7% of the original probes will be lost 

 mdat = QCfilter(mdat,qcinfo = qc,outCpG = discard)

# distribution of beta valus after background correction and probe filtering
 beta <- getB(mdat)
anno = rowData(mdat)   # get the EPIC Annotation data


beta1 = beta[anno$Infinium_Design_Type == "I", ]
beta2 = beta[anno$Infinium_Design_Type == "II", ]



p1 = multifreqpoly(beta,main="All probes",xlab="Beta value")

p2 = multifreqpoly(beta1,main="Infinium I probes",xlab="Beta value")

p3 = multifreqpoly(beta2,main="Infinium II probes",xlab="Beta value")
 

png("multifreq_plots_beta_values_background_corrected.png",height=6,width=18,units = "in",res = 600,type = "cairo")

print(grid.arrange(p1, p2, p3, nrow = 1))
dev.off()

# mdat<-norm.quantile(mdat, method="quantile1")

# Normalization using SWAN: takes into account the different distributions of the type I and type II probes
# and type II probe bias
# see minfi function and Maksimovic's paper (2012)

# needs to load full data set as methylSet object (not methDataSet as before)
sheet$Basename <- file.path(baseDir, sub("_.*", "", sheet$Sample_Name),sheet$Sample_Name)
minfi =  read.metharray.exp(targets = sheet)
minfi_mSet = preprocessRaw(minfi)

minfi_mSet = QCfilter(minfi_mSet,qcinfo = qc,outCpG = discard)

myNorm = preprocessSWAN(mSet=minfi_mSet,rgSet = minfi)

# get EPIC annotation data again from mdat object, but I need to filter it first so that mdata and myNorm have the same size

anno = rowData(mdat)   # get the EPIC Annotation data

differences = setdiff(mdat@NAMES,featureNames(myNorm))
anno = anno[which(rownames(anno) %notin% differences),]


# calculating beta values and plotting again
beta = getBeta(myNorm)
# anno and beta don't have probes in same order!! reordering
anno = anno[match(rownames(beta), rownames(anno)), ]

beta1 = beta[anno$Infinium_Design_Type == "I", ]
beta2 = beta[anno$Infinium_Design_Type == "II", ]



p1 = multifreqpoly(beta,main="All probes",xlab="Beta value")

p2 = multifreqpoly(beta1,main="Infinium I probes",xlab="Beta value")

p3 = multifreqpoly(beta2,main="Infinium II probes",xlab="Beta value")
 

png("multifreq_plots_beta_values_QN_PTBadjusted.png",height=6,width=18,units = "in",res = 600,type = "cairo")

print(grid.arrange(p1, p2, p3, nrow = 1))
dev.off()

##### none of these give better results than Matthias' method

```
When type I and type II probes are plotted separately the difference in modes between type I and II probes can be appreciated. But when all probes are plotted together (Fig 4 top panels), the multidensity plot obscures these differences, while they remain readily apparent in the multifreqpoly plot. In addition, the multidensity plots appear to suggest that probes range in value from <0 to >1, whereas multifreqpoly correctly show the range from 0 to 1.

The background is estimated using negative control positions, assuming a normal distribution. Calculations are performed on the original (non-log) scale.

A RatioSet object is a class designed to store Beta values and/or M values instead of the methylated and unmethylated signals. An optional copy number matrix, CN, the sum of the methylated and unmethylated signals, can be also stored. Mapping a MethylSet to a RatioSet may be irreversible, i.e. one cannot be guaranteed to retrieve the methylated and unmethylated signals from a RatioSet. A RatioSet can be created with the function ratioConvert:

```{r ratioset}

# The RGChannelSet stores also a manifest object that contains the probe design information of the array:

manifest <- getManifest(rgSet)
manifest

head(getProbeInfo(manifest))


RSet <- ratioConvert(mraw, what = "both", keepCN = TRUE)
RSet

  # Preprocessing
  # Method: Raw (no normalization or bg correction)
  # minfi version: 1.18.6
  # Manifest version: 0.3.0


# The functions getBeta, getM and getCN return respectively the Beta value matrix, M value matrix and the Copy Number matrix.

beta <- getBeta(RSet)

anno=getAnnotation(RSet)   # get the EPIC Annotation data

beta1=beta[anno$Type=="I",]
beta2=beta[anno$Type=="II",]


jpeg("/Users/Marta/Documents/WTCHG/DPhil/Data/Results/Methylation/multifreq_plots_beta_values_ratioset.jpg",height=900,width=600)

par(mfrow=c(3,1),mar=c(5, 5, 2, 5), xpd=TRUE)

multifreqpoly(beta,main="Multifreqpoly",xlab="Beta value")

multifreqpoly(beta1,main="Multifreqpoly: Infinium I",xlab="Beta value")

multifreqpoly(beta2,main="Multifreqpoly: Infinium II",xlab="Beta value")

dev.off()

# we see that the frequencies of the beta values haven't changed 

# same with M values:

mval<-getM(RSet)


mval1=mval[anno$Type=="I",]
mval2=mval[anno$Type=="II",]

# I can't plot directly, because there are NaN and Inf values (that are tried to be used by the function to set the min x value)


mval[is.infinite(mval)] <- NA
mval <- na.omit(mval)

mval1[is.infinite(mval1)] <- NA
mval1 <- na.omit(mval1)

mval2[is.infinite(mval2)] <- NA
mval2 <- na.omit(mval2)

jpeg("/Users/Marta/Documents/WTCHG/DPhil/Data/Results/Methylation/multifreq_plots_M_values.jpg",height=900,width=600)

par(mfrow=c(3,1),mar=c(5, 5, 2, 5), xpd=TRUE)

multifreqpoly(mval,main="Multifreqpoly",xlab="M value")

multifreqpoly(mval1,main="Multifreqpoly: Infinium I",xlab="M value")

multifreqpoly(mval2,main="Multifreqpoly: Infinium II",xlab="M value")

dev.off()

# remember that all that has been done doesn't include any normalization


```
The function mapToGenome applied to a RatioSet object will add genomic coordinates to each probe together with some additional annotation information. The output object is a GenomicRatioSet (class holding M or/and Beta values together with associated genomic coordinates). 
```{r genomicratioset}


# It is possible to merge the manifest object with the genomic locations by setting the option mergeManifest to TRUE.

GRset <- mapToGenome(RSet,mergeManifest=TRUE)
GRset



# Note that the GenomicRatioSet extends the class SummarizedExperiment. Here are the main accessors functions to access the data:

beta <- getBeta(GRset)
M <- getM(GRset)
CN <- getCN(GRset)

sampleNames <- sampleNames(GRset)
probeNames <- featureNames(GRset)
pheno <- pData(GRset)


# To return the probe locations as a GenomicRanges objects, one can use the accessor granges:

gr <- granges(GRset)
head(gr, n=3)

# To access the full annotation, one can use the command getAnnotation:


annotation <- getAnnotation(GRset)
names(annotation)


# It has been previously reported than about 5% of the probes on the EPIC array co-hybridize
# to alternate genomic sequences, therefore potentially generating spurious signals (Pidsley et al., 2016).
# Drop those probes.

dim(GRset)
GRset <- GRset[which(probeNames %notin% as.character(cross_reactive_EPIC[,1])),] # keep only probes that are not cross reactive
dim(GRset)
```

minfi provides a simple quality control plot that uses the log median intensity in both the methylated (M) and unmethylated (U) channels.When plotting these two medians against each other, it has been observed that good samples cluster together, while failed samples tend to separate and have lower median intensities. In order to obtain the methylated and unmethylated signals, we need to convert the RGChannelSet to an object containing the methylated and unmethylated signals using the function preprocessRaw. It takes as input a RGChannelSet and converts the red and green intensities to methylated and unmethylated signals according to the special 450K probe design, and returns the converted signals in a new object of class MethylSet. It does not perform any normalization.

```{r QC_minfi}
# The accessors getMeth and getUnmeth can be used to get the methylated and unmethylated intensities matrices:

head(getMeth(mraw)[,1:3])
head(getUnmeth(mraw)[,1:3])

# The functions getQC and plotQC are designed to extract and plot the quality control information from the MethylSet:

qc <- getQC(mraw)
head(qc)

par(mfrow=c(1,1)) 
plotQC(qc)

# the function addQC applied to the MethylSet will add the QC information to the phenotype data.

# To further explore the quality of the samples, it is useful to look at the Beta value densities of the samples, with the option to color 
# the densities by sampleID. Already done previously with frequencies.

# densityPlot(mraw, sampGroups = pD$sample)

# or density bean plots:

densityBeanPlot(mraw, sampGroups = pD$sample)

 # Control probes plot

# The 450k array contains several internal control probes that can be used to assess the quality control of different sample preparation 
# steps (bisulfite conversion, hybridization, etc.). The values of these control probes are stored in the initial RGChannelSet and can be 
# plotted by using the function controlStripPlot and by specifying the control probe type:

controlStripPlot(rgSet, controls="BISULFITE CONVERSION II") # not working?

# qcReport(rgSet, pdf= "qcReport.pdf")   # it's not plotting properly


```
The function preprocessQuantile implements stratified quantile normalization preprocessing for Illumina methylation microarrays.
```{r norm}

# normalize the data using Quantiles; this results in a GenomicRatioSet object 
gset.quantile <- preprocessQuantile(rgSet)

# another function, more apt for data with expected large scale differences:

gset.funnorm <- preprocessFunnorm(rgSet)

# visualise what the data looks like before and after normalisation 
jpeg("/Users/Marta/Documents/WTCHG/DPhil/Data/Results/Methylation/before_after_quantile_normalization.jpg",height=700,width=1400)
par(mfrow=c(2,1),mar=c(5, 5, 2, 5), xpd=TRUE)
densityPlot(rgSet, sampGroups=pD$sample,main="Raw", legend=FALSE) 
legend("topright", legend = levels(as.factor(pD$sample)), text.col=brewer.pal(8,"Dark2"),inset=c(-0.07,0))

densityPlot(getBeta(gset.quantile), sampGroups=pD$sample, main="Normalized", legend=FALSE)
legend("topright", legend = levels(as.factor(pD$sample)), text.col=brewer.pal(8,"Dark2"),inset=c(-0.07,0))

dev.off()

jpeg("/Users/Marta/Documents/WTCHG/DPhil/Data/Results/Methylation/before_after_funnorm_normalization.jpg",height=700,width=1400)
par(mfrow=c(2,1),mar=c(5, 5, 2, 5), xpd=TRUE)
densityPlot(rgSet, sampGroups=pD$sample,main="Raw", legend=FALSE) 
legend("topright", legend = levels(as.factor(pD$sample)), text.col=brewer.pal(8,"Dark2"),inset=c(-0.07,0))

densityPlot(getBeta(gset.funnorm), sampGroups=pD$sample, main="Normalized", legend=FALSE)
legend("topright", legend = levels(as.factor(pD$sample)), text.col=brewer.pal(8,"Dark2"),inset=c(-0.07,0))

dev.off()


# MDS plots
library(limma)
colours37 = c("#466791","#60bf37","#953ada","#4fbe6c","#ce49d3","#a7b43d","#5a51dc","#d49f36","#552095",
             "#507f2d","#db37aa","#84b67c","#a06fda","#df462a","#5b83db","#c76c2d","#4f49a3","#82702d",
             "#dd6bbb","#334c22","#d83979","#55baad","#dc4555","#62aad3","#8c3025","#417d61","#862977",
             "#bba672","#403367","#da8a6d","#a79cd4","#71482c","#c689d0","#6b2940","#d593a7","#895c8b",
             "#bd5975") # larger selection of colours

pal <- brewer.pal(5,"Dark2")

jpeg("/Users/Marta/Documents/WTCHG/DPhil/Data/Results/Methylation/mds_common_samples_stages_quantilenorm.jpg",height=700,width=1400)

par(mfrow=c(1,2),mar=c(5, 5, 2, 5), xpd=TRUE)

plotMDS(getM(gset.quantile), top=1000, gene.selection="common", col=colours37[as.factor(pD$stage)])
legend("topright", legend=levels(as.factor(pD$stage)), text.col=colours37, bg="white", cex=0.7)


plotMDS(getM(gset.quantile), top=1000, gene.selection="common", col=pal[as.factor(pD$sample)])
legend("topright", legend=levels(as.factor(pD$sample)), text.col=pal, bg="white", cex=0.7)

dev.off()

jpeg("/Users/Marta/Documents/WTCHG/DPhil/Data/Results/Methylation/mds_pairwise_samples_stages_quantilenorm.jpg",height=700,width=1400)

par(mfrow=c(1,2),mar=c(5, 5, 2, 5), xpd=TRUE)

plotMDS(getM(gset.quantile), top=1000, gene.selection="pairwise", col=colours37[as.factor(pD$stage)])
legend("topright", legend=levels(as.factor(pD$stage)), text.col=colours37, bg="white", cex=0.7)


plotMDS(getM(gset.quantile), top=1000, gene.selection="pairwise", col=pal[as.factor(pD$sample)])
legend("topright", legend=levels(as.factor(pD$sample)), text.col=pal, bg="white", cex=0.7)
dev.off()


# funnorm

jpeg("/Users/Marta/Documents/WTCHG/DPhil/Data/Results/Methylation/mds_common_samples_stages_funnorm.jpg",height=700,width=1400)
par(mfrow=c(1,2),mar=c(5, 5, 2, 5), xpd=TRUE)

plotMDS(getM(gset.funnorm), top=1000, gene.selection="common", col=colours37[as.factor(pD$stage)])
legend("topright", legend=levels(as.factor(pD$stage)), text.col=colours37, bg="white", cex=0.7)

plotMDS(getM(gset.funnorm), top=1000, gene.selection="common", col=pal[as.factor(pD$sample)])
legend("topright", legend=levels(as.factor(pD$sample)), text.col=pal, bg="white", cex=0.7)
dev.off()

jpeg("/Users/Marta/Documents/WTCHG/DPhil/Data/Results/Methylation/mds_pairwise_samples_stages_funnorm.jpg",height=700,width=1400)
par(mfrow=c(1,2),mar=c(5, 5, 2, 5), xpd=TRUE)

plotMDS(getM(gset.funnorm), top=1000, gene.selection="pairwise", col=colours37[as.factor(pD$stage)])
legend("topright", legend=levels(as.factor(pD$stage)), text.col=colours37, bg="white", cex=0.7)

plotMDS(getM(gset.funnorm), top=1000, gene.selection="pairwise", col=pal[as.factor(pD$sample)])
legend("topright", legend=levels(as.factor(pD$sample)), text.col=pal, bg="white", cex=0.7)
dev.off()

# Examine higher dimensions to look at other sources of variation 

jpeg("/Users/Marta/Documents/WTCHG/DPhil/Data/Results/Methylation/mds_common_samples_stages_secondaryPCs.jpg",height=700,width=2100)
par(mfrow=c(1,3)) 

plotMDS(getM(gset.quantile), top=1000, gene.selection="common", col=colours37[as.factor(pD$stage)], dim=c(1,3))
legend("topright", legend=levels(as.factor(pD$stage)), text.col=colours37, cex=0.7, bg="white")

plotMDS(getM(gset.quantile), top=1000, gene.selection="common", col=colours37[as.factor(pD$stage)], dim=c(2,3))
legend("topright", legend=levels(as.factor(pD$stage)), text.col=colours37, cex=0.7, bg="white")

plotMDS(getM(gset.quantile), top=1000, gene.selection="common", col=colours37[as.factor(pD$stage)], dim=c(3,4))
legend("topright", legend=levels(as.factor(pD$stage)), text.col=colours37, cex=0.7, bg="white")

dev.off()

```

```{r violin}
# fix
# Violin Plots of methylation levels across the whole genome
library(vioplot)
vioplots_samples <- data.frame(matrix(nrow=nrow(rgSet),ncol=ncol(rgSet)))
colnames(vioplots_samples)=colnames(rgSet)

for(s in colnames(rgSet)){

  vioplots_samples[,s] = rgSet[,s]


}
vioplot(rgSet, names=pD$simple_id, col="gold")
title("Violin Plots of methylation levels across the whole genome")
```